optuna/multi_objective/samplers/_nsga2.py

```diff
@@ -0,0 +1,254 @@
+        self._rng = np.random.RandomState(seed)
+
+    def infer_relative_search_space(
+        self,
+        study: "multi_objective.study.MultiObjectiveStudy",
+        trial: "multi_objective.trial.FrozenMultiObjectiveTrial",
+    ) -> Dict[str, BaseDistribution]:
+        return {}
+
+    def sample_relative(
+        self,
+        study: "multi_objective.study.MultiObjectiveStudy",
+        trial: "multi_objective.trial.FrozenMultiObjectiveTrial",
+        search_space: Dict[str, BaseDistribution],
+    ) -> Dict[str, Any]:
+        parent_generation, population = self._collect_parent_population(study)
+        trial_id = trial._trial_id
+
+        if len(population) == 0:
+            generation = 0
+            study._storage.set_trial_system_attr(trial_id, _GENERATION_KEY, generation)
+        else:
+            p0 = self._select_parent(study, population)
+            p1 = self._select_parent(study, population)
+
+            generation = parent_generation + 1
+            study._storage.set_trial_system_attr(trial_id, _GENERATION_KEY, generation)
+            study._storage.set_trial_system_attr(trial_id, _PARENTS_KEY, [p0.number, p1.number])
+
+        return {}
+
+    def sample_independent(
+        self,
+        study: "multi_objective.study.MultiObjectiveStudy",
+        trial: "multi_objective.trial.FrozenMultiObjectiveTrial",
+        param_name: str,
+        param_distribution: BaseDistribution,
+    ) -> Any:
+        if _PARENTS_KEY not in trial.system_attrs:
+            return self._random_sampler.sample_independent(
+                study, trial, param_name, param_distribution
+            )
+
+        trials = study.get_trials(deepcopy=False)
+        p0, p1 = trial.system_attrs[_PARENTS_KEY]
+
+        param = trials[p0].params.get(param_name, None)
+        if param is None or self._rng.rand() < self._crossover_prob:
+            param = trials[p1].params.get(param_name, None)
+
+        if param is None or self._rng.rand() < self._mutation_prob:
+            return self._random_sampler.sample_independent(
+                study, trial, param_name, param_distribution
+            )
+
+        return param
+
+    def _collect_parent_population(
+        self, study: "multi_objective.study.MultiObjectiveStudy"
+    ) -> Tuple[int, List["multi_objective.trial.FrozenMultiObjectiveTrial"]]:
+        # TODO(ohta): Optimize this method.
+
+        trials = study.get_trials(deepcopy=False)
+        parent_population = []  # type: List[multi_objective.trial.FrozenMultiObjectiveTrial]
+        parent_generation = -1
+        for generation in itertools.count():
+            population = [
+                t
+                for t in trials
+                if (
+                    t.system_attrs.get(_GENERATION_KEY, 0) == generation
+                    and t.state == optuna.trial.TrialState.COMPLETE
+                )
+            ]
+            if len(population) < self._population_size:
+                break
```

Comment:
Left the comment to the source code (https://github.com/optuna/optuna/pull/1163/commits/0e0bca869a98d2a38fd2e31fed574e0de23d7c54) as it would be beneficial for future readers.